In [1]:
import pandas as pd
import numpy as np

# Загружаем данные
old_data = pd.read_csv('music_old.csv') 
new_data = pd.read_csv('music_new.csv') 

In [2]:
# Колонка для расчёта PSI
col = 'loudness'

In [3]:
# Количество бинов для разбиения
n_bins = 20
quantiles = np.linspace(0, 1, n_bins + 1)

# Границы бинов по старым данным
bin_edges = old_data[col].quantile(quantiles).values

In [4]:
# Считаем распределение по бинам для старых данных
old_counts, _ = np.histogram(old_data[col], bins=bin_edges)
old_percents = old_counts / old_counts.sum()

# Считаем распределение по бинам для новых данных
new_counts, _ = np.histogram(new_data[col], bins=bin_edges)
new_percents = new_counts / new_counts.sum()

In [5]:
# Чтобы избежать деления на ноль
epsilon = 1e-8
old_percents = np.clip(old_percents, epsilon, 1)
new_percents = np.clip(new_percents, epsilon, 1)

In [6]:
# Расчёт PSI
psi_values = (old_percents - new_percents) * np.log(old_percents / new_percents)
psi = np.sum(psi_values)

print(f"Population Stability Index (PSI) for column '{col}': {psi:.6f}")

Population Stability Index (PSI) for column 'loudness': 0.235367


Наше значение (PSI ≈ 0.24) превышает порог в 0.2 и означает наличие расхождения между новыми данными и эталоном.